v2/blockstore/readonly.go

```diff
@@ -303,11 +313,26 @@ func (b *ReadOnly) PutMany([]blocks.Block) error {
 	panic("called write method on a read-only blockstore")
 }
 
-// AllKeysChan returns the list of keys in the CAR.
+// WithAsyncErrorHandler returns a context with async error handling set to the given errHandler.
+// Any errors that occur during asynchronous operations of AllKeysChan will be passed to the given
+// handler.
+func WithAsyncErrorHandler(ctx context.Context, errHandler func(error)) context.Context {
+	return context.WithValue(ctx, asyncErrHandlerKey, errHandler)
+}
+
+// AllKeysChan returns the list of keys in the CAR data payload.
+// If the ctx is constructed using WithAsyncErrorHandler any errors that occur during asynchronous
+// retrieval of CIDs will be passed to the error handler function set in context.
+// Otherwise, errors will terminate the asynchronous operation silently.
+//
+// See WithAsyncErrorHandler
 func (b *ReadOnly) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
 	// We release the lock when the channel-sending goroutine stops.
 	b.mu.RLock()
 
+	// Get async error handler from the context.
+	errHandler := asyncErrHandlerFrom(ctx)
```

Comment:
A drive-by comment: since all uses of errHandler are immediately followed by a return you could just get the error handler when you need it instead of pre-emptively here. I would probably rework `asyncErrHandlerFrom` into `maybeReportError(ctx context.Context, err error)` which does the context value unpacking and calls any discovered function. That also removes the need for `noopErrHandler` since `maybeReportError` would just return if no error handler was found in the context.